### PR TITLE
Add SOCKS capability to Lantern

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

File: MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        if (analyticsTrackingId != null && !enabled) {
+            trackStartSession(context, analyticsTrackingId);
+        }
+        return result;
+    }
+
+    private static go.lantern.Lantern.StartResult enable(String configDir, int timeoutMillis) {
+        try {
+            go.lantern.Lantern.StartResult result = go.lantern.Lantern.Start(configDir, timeoutMillis);
+            String addr = result.getHTTPAddr();
```
Comment (atavism) [inline MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java]: Does ``getHTTPAddr()`` always return a string that can be split on ':' now, assuming we get a valid result?

File: MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            String host = addr.split(":")[0];
+            String port = addr.split(":")[1];
+            System.setProperty("http.proxyHost", host);
+            System.setProperty("http.proxyPort", port);
+            System.setProperty("https.proxyHost", host);
+            System.setProperty("https.proxyPort", port);
+            return result;
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to start Lantern: " + e.getMessage(), e);
```
Comment (atavism) [inline MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java]: I wonder if throwing a ``RuntimeException`` here is too severe. It means Lantern could not be initialized, and there is nothing the client can do to recover. But applications that integrate the SDK may want the option to continue running regardless.

File: MobileSDK/sdk/build.gradle [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        versionCode 1
+        versionName "1.0"
+    }
+    buildTypes {
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
+    }
+}
```
Comment (atavism) [inline MobileSDK/sdk/build.gradle]: Are we going to add back all the publishing options/commands for uploading artifacts to Maven Central in a separate PR, or are we getting rid of all that for some reason? The whole point of a separate SDK project was to turn Lantern into an Android library our own apps ("Lantern Android", FireTweet) and third party apps could integrate easily http://lantern.readthedocs.org/en/latest/ Just clarifying.

**Comment by atavism** [comment]
We probably want to remove all the ``.class`` files from this PR and add a .gitignore *.class entry? Actually, we shouldn't include the ``app/build`` directory at all

File: MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java (hunk 1 above)
Comment (oxtoacart) [inline MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java]: For IPv4 it does, but for IPv6 we actually need to use the portion after the last colon as the port.  I'll fix that.

File: MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java (hunk 2 above)
Comment (oxtoacart) [inline MobileSDK/sdk/src/main/java/org/lantern/mobilesdk/Lantern.java]: They can do that by catching the exception.  I'll make it a checked exception to be clear.

File: MobileSDK/sdk/build.gradle (hunk 3 above)
Comment (oxtoacart) [inline MobileSDK/sdk/build.gradle]: Still need to add that back in.  I could use your help with that, as I don't understand it very well.

**Comment by oxtoacart** [comment]
@atavism Sorry about those build outputs, that's fixed.  I've addressed your other two comments as well.  I also noticed that I committed some .gradle directories, that's fixed in the final PR (valencia-mobile-full).

File: MobileSDK/sdk/build.gradle (hunk 3 above)
Comment (atavism) [inline MobileSDK/sdk/build.gradle]: Oh sure =P Let's just do that in a separate PR as this one is already massive!
